src/common/object/node/object.hpp

```diff
@@ -0,0 +1,157 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "common/collection.hpp"
+#include "methods.hpp"
+#include "napi.h"
+#include "subscriber.hpp"
+
+namespace realm {
+namespace common {
+
+template <typename GetterSetter>
+class JavascriptObject {
+   private:
+    Napi::Env context;
+    Napi::Reference<Napi::Object> ref_object;
+    ObjectObserver *observer = nullptr;
+    IOCollection *collection = nullptr;
+    std::vector<std::string> keys;
+
+    template <class T>
+    using ObjectAPI = realm::js::Object<T>;
+    using Property = realm::js::PropertyAttributes;
+
+    bool remove_key(std::string &key) {
+        int index = -1;
+        for (auto const &_key : keys) {
+            index++;
+            if (key == _key) {
+                keys.erase(keys.begin() + index);
+                return true;
+            }
+        }
+        return false;
+    }
+
+   public:
+    JavascriptObject(Napi::Env _ctx, std::string name = "js_object")
+        : context{_ctx} {
+        ref_object =
+            Napi::Reference<Napi::Object>::New(Napi::Object::New(context));
+    }
+
+    void set_collection(IOCollection *_collection) { collection = _collection; }
+
+    void set_observer(ObjectObserver *_observer) { observer = _observer; }
+
+    template <class VM, void callback(method::Arguments)>
+    void add_method(std::string &&name) {
+        auto object = get();
+
+        auto method = [=](const auto &info) {
+            callback({info.Env(), observer, collection, info.Length(),
```

Comment:
> Do we need a HANDLESCOPE here?
No, because the JSObject class shouldn't be not concerned about the current scope that the user (the dev) want to use.